### PR TITLE
[now-build-utils] Throw error on Next.js pages/api and api/

### LINKED ISSUE #2887: Throw error if "api" + "pages/api" is used in @now/next
Stumbled upon something interesting with [now's zero-config api](https://zeit.co/guides/upgrade-to-zero-configuration#with-zero-configuration) and [Next's API routes](https://nextjs.org/docs#api-routes). I had a Next 9 project with the following folder structure (truncated for readability):

```
.
├── api
│   └── index.js <--- API Utility Helper
├── pages
│   ├── api
│   └── users.js <--- Next API Route
└── static
```

I was not aware of the zero-config API feature and was using the top-level `/api` folder to organize api-related utilities. I ran the app locally with `now dev` and everything worked as expected. The problem, however, came when I deployed this with `now`. The calls to `/api/users` no longer succeeded and instead would return a `404` response code. Strangely though, the expected data was also returned.

After reading through the docs and learning about now's zero-config API feature, I realized that  the top-level `api` folder was causing a conflict with Next's API routes. The solution seems to be use one or the other and not both, however I still don't understand what causes the `404` response with data.

I have an [example repo](https://github.com/adamgruber/now-api-test) deployed to `https://now-api-test.adamgruber.now.sh` with two endpoints. One using the zero-config now API and one using Next API routes. When hitting these endpoints, I get the following responses:

```
❯ http https://now-api-test.adamgruber.now.sh/api/next
HTTP/1.1 404 Not Found
Connection: keep-alive
Content-Type: application/json; charset=utf-8
cache-control: public, max-age=0, must-revalidate
content-length: 13
date: Tue, 27 Aug 2019 14:26:22 GMT
server: now
strict-transport-security: max-age=63072000
x-now-cache: MISS
x-now-id: iad1:sfo1:pdrjv-1566915981899-88066e99a277
x-now-trace: iad1

{
    "next": true
}

❯ http https://now-api-test.adamgruber.now.sh/api/now
HTTP/1.1 200 OK
Connection: keep-alive
Content-Type: application/json; charset=utf-8
cache-control: public, max-age=0, must-revalidate
content-length: 12
date: Tue, 27 Aug 2019 14:27:37 GMT
etag: W/"c-yfKwHQSNx6jM3PPrBNeiUcNYTfQ"
server: now
strict-transport-security: max-age=63072000
x-now-cache: MISS
x-now-id: iad1:sfo1:pdrjv-1566916057438-ebb579b76baa
x-now-trace: iad1

{
    "now": true
}

```

I'm really curious to know what causes this odd behavior and why it didn't manifest when using `now dev`.

### PR DESCRIPTION
**Pull request by AndyBitz**
There can be an unexpected outcome when using Next.js pages/api and api/ at the same time, because of this we want to error when we see that both exist.

Fixes https://github.com/zeit/now/issues/2887

### CODE REVIEW

**Review by Timer — DISMISSED** [review]
This seems like an arbitrary constraint that is already breaking my local usage of these feature(s).

Are you sure this is what we want? IMO it should be a warning.

**Comment by AndyBitz** [comment]
@Timer is there a reason why you use both at the same time instead of just one?

**Comment by Timer** [comment]
Next.js doesn't support non-JS APIs -- I'm already using Next.js w/ a combination of Golang APIs.

Maybe we can only error for `.js` or `.ts` files in top-level `api/`?

**Comment by AndyBitz** [comment]
@Timer but the routes could still conflict:

- `api/user.go`
- `pages/api/user.js`
- `pages/api/[endpoint].js`
- `api/[endpoint].go`

**Comment by Timer** [comment]
I'm aware the routes can conflict -- but completely disabling zero-config `api/` seems like a big handicap.

If you move forward with this change, please release it as a **breaking change**, because it is. 😄

**Comment by styfle** [comment]
@AndyBitz That is a good point, unless the user is planning to use define routes. What about checking if `routes` is defined in now.json and throwing if it is missing?

File: packages/now-build-utils/src/detect-builders.ts [hunk 1]
```diff
@@ -107,6 +107,31 @@ async function detectApiBuilders(files: string[]): Promise<Builder[]> {
...
+  builders: Builder[]
+): Promise<ErrorResponse | null> {
+  // For Next.js
+  if (builders.some(builder => builder.use.startsWith('@now/next'))) {
+    const hasApiPages = files.some(file => file.startsWith('pages/api/'));
+    const hasApiBuilders = builders.some(builder =>
+      builder.src.startsWith('api/')
+    );
+
+    if (hasApiPages && hasApiBuilders) {
```
Comment (styfle) [inline packages/now-build-utils/src/detect-builders.ts:123]: Can we pass `nowJson.routes` into this function?

Then we could add `const hasRoutes = routes && routes.length > 0`.

```suggestion
    if (hasApiPages && hasApiBuilders && !hasRoutes) {
```

**Comment by AndyBitz** [comment]
@styfle I don't think that this is an option, since we don't have access to `now.json` here.

And even if they were defined, zero-config and Next.js will still apply their own routes, so it's not really a guarantee that the user knows about what will happen, or if the routes from `now.json` have any relation to the API.

@Timer agreed, but we'll always use the latest version of this function in our API to determine the builders and routes for the user, so it's not possible to differentiate.

cc @leo what do you think? Maybe a warning would already be enough, as @Timer suggested?

**Comment by TooTallNate** [comment]
I'm 👍 for a warning.

However, I am curious @Timer, why not place the `.js` files in the root `api` directory?

**Comment by leo** [comment]
I don't see a reason for not showing an error for new projects.

Sure, Next.js only supports JS APIs, but we're not asking for erroring if you have `pages/api` – only if you have it AND `api`.

**Review by styfle — APPROVED** [review]
I thought about this again and I agree with @leo 

If someone wants to mix `.js` and `.go` APIs, they can use the root `/api` folder instead of `/pages/api`. Let's always error like you wrote it 👍

**Pull request merged**